### PR TITLE
AngularJS: compatibility with strictFunctionTypes

### LINKED ISSUE #21160: AngularJS typings with TypeScript's 2.6 strictFunctionTypes don't allow extending IScope
Creating a directive with a customized scope object is not allowed when `strictFunctionTypes` (and therefore `strict` is enabled).

This code works with `strictFunctionTypes` disabled:  

```ts
import angular from 'angular'

const m = angular.module('exampleModule', [])

interface IMyScope extends angular.IScope {
  myScopeProperty: boolean,
}

m.directive('directive', () => ({
  link(scope: IMyScope) {
    console.log(scope.myScopeProperty)
    return
  }
}))
```

But when enabled it throws:

> Property 'includes' is missing in type '() => { link(scope: IMyScope): void; }'
 
This is because `IMyScope` is not assignable to `angular.IScope`:

```ts
  link(scope: angular.IScope) {
    const x: IMyScope = scope // Type 'IScope' is not assignable to type 'IMyScope'.
                              // Property 'myScopeProperty' is missing in type 'IScope'.
  }
```

I don't think there is a way to implement this with the new strict mode without a type assertion. Perhaps the right solution is to have `IDirective` have a generic type parameter for your scope type?

----------

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @.... @diegovilar @thorn0 @calebstdenis @leonard-thieu

### PR DESCRIPTION
**Pull request by thorn0**
Fixes #21160 

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@thorn0 Please fix the failures indicated in the Travis CI log.

**Review by leonard-thieu — APPROVED** [review]

**Comment by RyanCavanaugh** [comment]
Approved by a listed owner. PR ready to merge pending express review by a maintainer.

**Comment by aciccarello** [comment]
Is there any way to update these type definitions so that specifying a TScope is optional?

See ngParty/ng-metadata#214:

> Generic type 'IDirectiveCompileFn<TScope>' requires 1 type argument(s).

**Pull request merged**
